src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/DoNotUseOutAttributeStringPInvokeParametersTests.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Testing;
+using Xunit;
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.InteropServices.DoNotUseOutAttributeStringPinvokeParametersAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.InteropServices.DoNotUseOutAttributeStringPinvokeParametersAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+namespace Microsoft.NetCore.Analyzers.InteropServices.UnitTests
+{
+    public class DoNotUseOutAttributeStringPInvokeParametersAnalyzerTests
+    {
+        [Fact]
+        public async Task StringByReference_NoDiagnostics_Diagnostics_CS()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+
+public class C
+{
+    [DllImport(""user32.dll"", CharSet=CharSet.Unicode)]
+    private static extern void Method1(out string s); // OK
+
+    [DllImport(""user32.dll"", CharSet=CharSet.Unicode)]
+    private static extern void Method2([In] [Out] ref string s); // OK
+
+    [DllImport(""user32.dll"", CharSet=CharSet.Unicode)]
+    private static extern void Method3([Out] out string s); // OK
+}
+";
+            await VerifyCS.VerifyAnalyzerAsync(source);
+        }
+
+        [Fact]
+        public async Task OutAttributeStringByValue_Diagnostics_CS()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+
+public class C
+{
+    [DllImport(""user32.dll"", CharSet=CharSet.Unicode)]
+    private static extern void Method1([Out] string {|#0:s|}); // Should not have [Out] string
```

Comment:
Can we write a code fixer for this diagnostic or did we feel that is not required? If not, then can the diagnostic message be made more actionable for users to indicate how to fix the issue?

Reply:
Let me know if you need help in writing a code fix for it.

Reply:
We had gone with just having an analyzer and a fixer not being required. I made the message a little more actionable to lead down the path of:
- Check if marshalling data out from the callee back to the caller is actually required
  - If yes, pass by reference
  - If no, remove the attribute